servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/AsyncContextDisableNewTest.java

```diff
@@ -0,0 +1,116 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.context.api.ContextMap;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.servicetalk.context.api.ContextMap.Key.newKey;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+class AsyncContextDisableNewTest {
+    private static final ContextMap.Key<String> K1 = newKey("k1", String.class);
+
+    @Test
+    void testDisableAsyncContext() throws Exception {
+        synchronized (K1) { // prevent parallel execution because these tests rely upon static state
+            Executor executor = Executors.newCachedThreadExecutor();
+            Executor executor2 = null;
+            try {
+                // Test that AsyncContext is enabled first.
+                String expectedValue = "foo";
+                AsyncContext.put(K1, expectedValue);
+                assertEquals(expectedValue, executor.submit(() -> AsyncContext.get(K1)).toFuture().get());
+                AtomicReference<String> actualValue = new AtomicReference<>();
+                Publisher.from(1, 2).publishOn(executor).beforeOnComplete(() -> actualValue.set(AsyncContext.get(K1)))
+                        .toFuture().get();
+                assertEquals(expectedValue, actualValue.get());
+                actualValue.set(null);
+                Single.succeeded(1).publishOn(executor).beforeOnSuccess(i -> actualValue.set(AsyncContext.get(K1)))
+                        .toFuture().get();
+                assertEquals(expectedValue, actualValue.get());
+                actualValue.set(null);
+                Completable.completed().publishOn(executor)
+                        .beforeOnComplete(() -> actualValue.set(AsyncContext.get(K1))).toFuture().get();
+                assertEquals(expectedValue, actualValue.get());
+                actualValue.set(null);
+
+                AsyncContext.disable();
+                try {
+                    // Create a new Executor after we have disabled AsyncContext so we can be sure that AsyncContext
+                    // won't be captured.
+                    executor2 = Executors.newCachedThreadExecutor();
+                    asyncContextPutExpectNoop(K1, expectedValue);
+                    assertNull(executor2.submit(() -> AsyncContext.get(K1)).toFuture().get());
+                } finally {
+                    AsyncContext.enable();
+                }
+            } finally {
+                if (executor2 != null) {
+                    executor2.closeAsync().toFuture().get();
+                }
+                executor.closeAsync().toFuture().get();
```

Comment:
I guess you'd also want to add `AsyncContext.enable();` if anything goes wrong in the outer try block.

Reply:
Not necessary, we disable it right before the `try` block. Exceptions outside this range won't affect `AsyncContext` state.